# Pandas 入門

[`pandas`](https://pandas.pydata.org/docs/index.html) は 様々な I/O (入出力) 処理、高機能なデータ処理の機能を提供する Python の外部ライブラリです。
データファイルの読み込み・書き込みや、複数のデータの結合などを可能にします。
まず、`pandas` の基本的な使い方について説明します。

## 基本的なデータ構造の概念

`pandas` で基本となるデータ構造 (クラス) は、`Series` と `DataFrame` です。
`Series` と `DataFrame` は、今までに学んだリストや辞書などのデータ構造と似ている部分もありますが、異なる部分もあります。
特に重要な概念について先に触れておきます。

### インデックス・カラム

`Series` は 1 次元のラベル付きベクトル (または配列) で、`DataFrame` は 2 次元のラベル付き行列 (または表形式のデータ)です。
ラベルを用いることで特定の要素、`DataFrame` であれば特定の行や列のデータを参照できます。

**インデックス** は、データを横に切った場合の *行* を識別するためのラベルです。
`Series` と `DataFrame` どちらでも、インデックスによってデータの位置が特定できます。

**カラム** は、データを縦に切った場合の *列* を識別するためのラベルです。
`DataFrame` では、カラム毎に異なるデータ型を持つことができ、それぞれが特定の種類のデータを表します。
例えば、`price` 列は価格の数値、`name` は商品名の文字列など。

#### データ型

**データ型** は、`DataFrame` のそれぞれのカラムのデータや `Series` のデータが数値、文字列、日付など何を表すかを定義します。
適切なデータ型を使用することが、データの分析・操作を効率的かつ正確に行うために重要です。

## Series, DataFrame の作成例

以下は `Series` の作成例です。


`size` 属性は 行数 x 列数、つまりの要素の数を返します。

In [1]:
from pandas import Series
Series([1, 2, 3])

0    1
1    2
2    3
dtype: int64

`0, 1, 2` というのが、各要素 (行) ごとについているラベル (インデックス) です。
このラベルを用いて、各要素 (行) を参照できます。

In [ ]:
data = Series([1, 2, 3])
data[0]

np.int64(1)

以下は `DataFrame` の作成例です。


In [3]:
from pandas import DataFrame
DataFrame({"item": ["apple", "melon"], "price": [100, 1000]})

item  price
0  apple    100
1  melon   1000

`item`, `price` はカラムのラベルを示しています。
以下は、ラベルを用いて特定の列 (カラム) を参照しています。


In [4]:
data = DataFrame({"item": ["apple", "melon"], "price": [100, 1000]})
data["item"]

0    apple
1    melon
Name: item, dtype: object

R を使ったことがある人はピンとくるかもしれませんが、DataFrame は R の data.frame 型と同じようなものです。 実際、pandas は R のデータ整理用のパッケージ、dplyr でできることの多くを実装しています。

## pandas のインポート

`pandas` のインポートは次のように行います。

In [5]:
import pandas

通常、`pandas` は `pd` という略称でインポートされることが多いです。

In [ ]:
import pandas as pd

`pd` という略称は、`pandas` を使用する際の一般的な慣例です。
これは多くの専門書やオンライン資料で一般的に見られます。
標準化された略称を使用することで、コードの可読性が向上し、他の開発者がコードを理解しやすくなります。

以降では、`pandas.<name>` を `pd.<name>` として表記します。

## Series の基礎

### Series の作成

`Series` はリストやタプルなどのシーケンス型から生成できます。

In [ ]:
Series([1, 2, 3])

0    1
1    2
2    3
dtype: int64

リストやタプルなどと異なる部分があることがわかります。
順に説明します。

### インデックス

左の `0, 1, 2` がこの `Series` インスタンスのインデックスです。
リストやタプルなどでもインデックスがありましたが、それとは少し異なります。
`Series` や `DataFrame` におけるインデックスは、ある行を表す名前です。
何も指定しない場合、先ほどのように、`0, 1, 2` という通し番号が振られます。
リストやタプルにおけるインデックスは通し番号であったのに対し、`Series` や `DataFrame` のインデックスは指定・変更することが可能です。

インデックスをあらかじめ指定するには、クラスを作成するときの引数にインデックスのシーケンスを加えます。

In [8]:
Series([1, 2, 3], index=["A", "B", "C"])

A    1
B    2
C    3
dtype: int64

インデックスは `index` 属性として呼び出し可能です。

In [ ]:
data = Series([1, 2, 3])
data.index


RangeIndex(start=0, stop=3, step=1)

`RangeIndex(start=0, stop=3, step=1)` は `range` 関数が返すイテレーターのようなもので、0 から 3 の直前 (2) までの値を順に返すイテレーターであるということを意味しています。

上書きをすることでインデックスを更新できます。

In [10]:
data.index = ["A", "B", "C"]
data

A    1
B    2
C    3
dtype: int64

データの長さとインデックスの長さ、つまり要素の数が一致していない場合はエラーになります。
以下のエラー文では、3 つの要素が求められているが、1 つしか得られなかったことでエラー (`ValueError`) になったと説明しています (コメントアウト `#` を削除して実行してください)。

In [11]:
data = Series([1, 2, 3])
# data.index = ["A"]

`Series` や `DataFrame` では、インデックスの値の重複を許します。
例えば、`Series([1, 2, 3], index=["A", "A", "B"])` はエラーにはなりません。
しかし、インデックスが重複すると思わぬエラーや意図せぬ挙動を引き起こす可能性があるため、インデックスはデータ内で一意にすることをおすすめします。

### Series の要素の取り出し

インデックスを指定することで、それに対応するデータの値を取り出すことができます。

In [12]:
data = Series([1, 2, 3], index=["A", "B", "C"])
data["A"]

np.int64(1)

`loc` プロパティというものを使っても値を参照することができます。
プロパティは、インスタンスから値を参照したり、逆に値を追加したりするときに使うための機能を指します。
プロパティの呼び出しは、属性やメソッドと同じようにドット `.` を用います。
プロパティの使い方は、プロパティがどのような機能を持っているかによって異なってきます。

例えば、`loc` プロパティは次のように使います。`loc("A")` ではないことに注意してください。

In [13]:
data.loc["A"]

np.int64(1)

また、`iloc` プロパティというものもあり、これはリストやタプルと同じように、通し番号によって要素を参照できます。
Python ではお約束通り、通し番号は `0` から始まります。

`iloc` プロパティの使い方は `loc` プロパティと似ています。

In [14]:
data.iloc[1]

np.int64(2)

また、負の値の通し番号も振られています。

In [15]:
data.iloc[-1]

np.int64(3)

### データ型

`dtype` はこの `Series` のデータ型を示しています。
これは、中に含まれているデータ全体の型を表しています。
基本的にはデータの型を指定する必要はなく、`Series` を作成すると自動的にデータの全体の型を推測してくれます。

In [16]:
Series([1, 2, 3])

0    1
1    2
2    3
dtype: int64

データは整数 (`int`) 型だったので、`dtype` は `int64` になりました。

`int` の後の数字はビット数で、64 ビットは -9223372036854775808 から 9223372036854775807 までの整数値を扱えます。
他に、`int8` や `int16`、`int32` などがあります。
基本的には、ビット数が多いものを使えば良いでしょう。

小数点型をデータに含めるとどうなるでしょうか。

In [17]:
Series([1.25, 2, 3])

0    1.25
1    2.00
2    3.00
dtype: float64

`dtype` は小数点のデータ型、`float64` になったことがわかります。

文字列をデータに加えてみます。

In [18]:
Series(["A", 1, 2])

0    A
1    1
2    2
dtype: object

## DataFrame の基礎

### DataFrame の作成

`DataFrame` は値がリストの辞書やリストのリストなど、いくつかの方法を使ってデータを作成できます。
しかし、データの作り方によってどのような `DataFrame` ができるかが異なるので注意が必要です。

辞書の場合、各キーのリストが各カラムに対応します。

In [19]:
DataFrame({"item": ["apple", "melon"], "price": [100, 1000]})

item  price
0  apple    100
1  melon   1000

リストが入れ子になっている場合、各リストのインデックスが各カラムに対応します。

In [20]:
data = DataFrame([["apple", 100], ["melon", 1000]])
data

0     1
0  apple   100
1  melon  1000

### インデックス・カラム

`DataFrame` では各インデックスは行、カラムは列に対応しています。
`Series` と同様、インデックスを `DataFrame` 作成時に指定することが可能です。

In [21]:
DataFrame({"item": ["apple", "melon"], "price": [100, 1000]}, index=["A", "B"])

item  price
A  apple    100
B  melon   1000

または、`index` 属性を上書きします。

In [22]:
data = DataFrame({"item": ["apple", "melon"], "price": [100, 1000]})
data.index = ["A", "B"]
data

item  price
A  apple    100
B  melon   1000

カラムも、インデックスと同様に、指定しない場合は自動で連番が振られます。
`columns` という引数名で指定可能です。
なお、辞書から作成した場合にはキーがカラム名として与えられます。

In [23]:
DataFrame([["apple", 100], ["melon", 1000]], columns=["item", "price"])

item  price
0  apple    100
1  melon   1000

`columns` 属性でカラム名を取得可能で、また上書きをすることもできます。

In [24]:
data = DataFrame({"item": ["apple", "melon"], "price": [100, 1000]})
data.columns

Index(['item', 'price'], dtype='object')

In [25]:
data.columns = ["I", "P"]
data

I     P
0  apple   100
1  melon  1000

### DataFrame 内のデータの参照

`DataFrame` は `Series` よりもデータの次元が 1 つ多いため、データの取り出し方が少し複雑になります。
まず、特定のカラムを参照する場合には、鍵括弧 `[]` を使います。
`Series` の時ではこれはインデックスでしたが、`DataFrame` ではカラムなのに注意してください。
また、列のデータが `Series` として返されることにも注意してください。

In [26]:
data = DataFrame({"item": ["apple", "melon"], "price": [100, 1000]})
data["price"]

0     100
1    1000
Name: price, dtype: int64

複数のカラムを指定する場合、鍵括弧内にカラム名の *リスト* を入れます。
リストを入れた場合には `DataFrame` が返ってきます。

In [27]:
data = DataFrame({
    "item": ["apple", "melon"],
    "price": [100, 1000],
    "sold": [10, 1],
})
data[["price", "sold"]]

price  sold
0    100    10
1   1000     1

複数の列を指定するときにリストで指定することに注意してください。
以下のようにすると、`price` と `solid` ではなく、`(price, solid)` というカラムを指定していることになります。

In [28]:
# data["price", "sold"]

`data["<name>"]` という指定ではなく、`data.name` という指定でもできます。
ただし、カラム名にスペースが入っている場合には後者の方法は使えません。

後者の方法を使うために、カラム名はスペースではなくアンダーバー `_` で単語を結合するのが良いでしょう。

In [29]:
data = DataFrame({
    "mean price": [100, 1000],
    "mean_price": [50, 500],
})
data

mean price  mean_price
0         100          50
1        1000         500

In [30]:
print(data["mean price"]) # OK
# print(data.mean price) # NG
print(data["mean_price"])  # OK
print(data.mean_price)  # OK

0     100
1    1000
Name: mean price, dtype: int64
0     50
1    500
Name: mean_price, dtype: int64
0     50
1    500
Name: mean_price, dtype: int64


縦ではなく横、つまりあるインデックスに対応する行のデータを取り出したい場合には、`loc` プロパティもしくは `iloc` プロパティを使います。
`loc` プロパティでインデックスを指定すると、それに対応する行の `Series` が返されます。

In [31]:
data = DataFrame({"item": ["apple", "melon"], "price": [100, 1000]}, index=["A", "B"])
data.loc["A"]

item     apple
price      100
Name: A, dtype: object

`DataFrame` から列を一つ取り出した場合、その列は `Series` オブジェクトです。

In [32]:
data = DataFrame({"item": ["apple", "melon"], "price": [100, 1000]}, index=["A", "B"])
print(type(data["price"]))

<class 'pandas.core.series.Series'>


複数列を取り出した場合は、`DataFrame` オブジェクトです。

In [33]:
print(type(data[["price", "item"]]))

<class 'pandas.core.frame.DataFrame'>


さらに、表形式データでいう、ある行のある列の要素だけを取り出したい場合には、次のように `loc` プロパティでインデックスを指定してからカラム名を指定します。

In [34]:
data.loc["A", "item"]

'apple'

### DataFrame のデータ型

`DataFrame` では、各データ型はカラムごとに定義されており、インスタンス作成時に自動でデータ型を類推してくれます。
また、各カラムのデータ型と、`DataFrame` 全体のデータ型は `dtypes` 属性 (複数形なのに注意!) で確認することができます。

In [35]:
data = DataFrame({"item": ["apple", "melon"], "price": [100, 1000]})
data.dtypes

item     object
price     int64
dtype: object

`Series` を `DataFrame` に変換するには、`to_frame` メソッドを用います。

In [36]:
series = Series([1, 2, 3], index =["A", "B", "C"])
df = series.to_frame()
print(type(df))
df

<class 'pandas.core.frame.DataFrame'>


0
A  1
B  2
C  3

## データの確認方法

### `head`, `tail`: データの一部を表示

データの行数、列数が多い場合、`DataFrame` や `Series` を確認すると、一部が省略された結果になります。
例えば、以下のようにカリフォルニア住宅データを読み込み、内容を確認してみます。


In [37]:
data = pd.read_csv("/content/sample_data/california_housing_test.csv")
data

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0       -122.05     37.37                27.0       3885.0           661.0   
1       -118.30     34.26                43.0       1510.0           310.0   
2       -117.81     33.78                27.0       3589.0           507.0   
3       -118.36     33.82                28.0         67.0            15.0   
4       -119.67     36.33                19.0       1241.0           244.0   
...         ...       ...                 ...          ...             ...   
2995    -119.86     34.42                23.0       1450.0           642.0   
2996    -118.14     34.06                27.0       5257.0          1082.0   
2997    -119.70     36.30                10.0        956.0           201.0   
2998    -117.12     34.10                40.0         96.0            14.0   
2999    -119.63     34.42                42.0       1765.0           263.0   

      population  households  median_income  median_house_value  
0         1537.0       606.0         6.6085            344700.0  
1          809.0       277.0         3.5990            176500.0  
2         1484.0       495.0         5.7934            270500.0  
3           49.0        11.0         6.1359            330000.0  
4          850.0       237.0         2.9375             81700.0  
...          ...         ...            ...                 ...  
2995      1258.0       607.0         1.1790            225000.0  
2996      3496.0      1036.0         3.3906            237200.0  
2997       693.0       220.0         2.2895             62000.0  
2998        46.0        14.0         3.2708            162500.0  
2999       753.0       260.0         8.5608            500001.0  

[3000 rows x 9 columns]

データの先頭から指定行表示する場合、`head` メソッドを用います。
`head` メソッドは、デフォルトでは先頭 5 行を返します。
表示行数は引数で指定可能です。

In [38]:
data.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.05     37.37                27.0       3885.0           661.0   
1    -118.30     34.26                43.0       1510.0           310.0   
2    -117.81     33.78                27.0       3589.0           507.0   
3    -118.36     33.82                28.0         67.0            15.0   
4    -119.67     36.33                19.0       1241.0           244.0   

   population  households  median_income  median_house_value  
0      1537.0       606.0         6.6085            344700.0  
1       809.0       277.0         3.5990            176500.0  
2      1484.0       495.0         5.7934            270500.0  
3        49.0        11.0         6.1359            330000.0  
4       850.0       237.0         2.9375             81700.0

In [39]:
# 最初の 1 行目のみ表示
data.head(1)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.05     37.37                27.0       3885.0           661.0   

   population  households  median_income  median_house_value  
0      1537.0       606.0         6.6085            344700.0

データの最後から指定行を表示する場合、`tail` メソッドを用います。
`tail` メソッドもデフォルトで最後の 5 行を返し、表示行数を引数で指定することも可能です。


In [40]:
data.tail()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2995    -119.86     34.42                23.0       1450.0           642.0   
2996    -118.14     34.06                27.0       5257.0          1082.0   
2997    -119.70     36.30                10.0        956.0           201.0   
2998    -117.12     34.10                40.0         96.0            14.0   
2999    -119.63     34.42                42.0       1765.0           263.0   

      population  households  median_income  median_house_value  
2995      1258.0       607.0         1.1790            225000.0  
2996      3496.0      1036.0         3.3906            237200.0  
2997       693.0       220.0         2.2895             62000.0  
2998        46.0        14.0         3.2708            162500.0  
2999       753.0       260.0         8.5608            500001.0

In [41]:
# 最後の 1 行のみ表示
data.tail(1)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2999    -119.63     34.42                42.0       1765.0           263.0   

      population  households  median_income  median_house_value  
2999       753.0       260.0         8.5608            500001.0

`pd.options.display.max_rows` を変更することで、省略なしに表示できる行数を変更することも可能です。
デフォルトでは、これは 60 で、`pd.options.display.max_rows = 1000` とすれば、1000 行まで省略なしで表示されるようにできます。
他の `pandas` のオプションに関しては、[ドキュメント](https://pandas.pydata.org/docs/user_guide/options.html)を見てください。

### `info`: データの概要情報を取得

`info` メソッドを使うと、各カラムのデータ型 (Dtype) だけでなく欠損値以外の値の数 (Non-Null Count) やデータ容量 (memory usage) などの情報を得ることができます。

In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           3000 non-null   float64
 1   latitude            3000 non-null   float64
 2   housing_median_age  3000 non-null   float64
 3   total_rooms         3000 non-null   float64
 4   total_bedrooms      3000 non-null   float64
 5   population          3000 non-null   float64
 6   households          3000 non-null   float64
 7   median_income       3000 non-null   float64
 8   median_house_value  3000 non-null   float64
dtypes: float64(9)
memory usage: 211.1 KB


### `describe`: 数値型データの記述統計量を取得

`describe` メソッドを使うと、各カラムの平均や標準偏差などの記述統計量を確認することができます。
数値型の列に対してのみ適用される点に注意してください。

In [43]:
data.describe()

longitude    latitude  housing_median_age   total_rooms  \
count  3000.000000  3000.00000         3000.000000   3000.000000   
mean   -119.589200    35.63539           28.845333   2599.578667   
std       1.994936     2.12967           12.555396   2155.593332   
min    -124.180000    32.56000            1.000000      6.000000   
25%    -121.810000    33.93000           18.000000   1401.000000   
50%    -118.485000    34.27000           29.000000   2106.000000   
75%    -118.020000    37.69000           37.000000   3129.000000   
max    -114.490000    41.92000           52.000000  30450.000000   

       total_bedrooms    population  households  median_income  \
count     3000.000000   3000.000000  3000.00000    3000.000000   
mean       529.950667   1402.798667   489.91200       3.807272   
std        415.654368   1030.543012   365.42271       1.854512   
min          2.000000      5.000000     2.00000       0.499900   
25%        291.000000    780.000000   273.00000       2.544000   
50%        437.000000   1155.000000   409.50000       3.487150   
75%        636.000000   1742.750000   597.25000       4.656475   
max       5419.000000  11935.000000  4930.00000      15.000100   

       median_house_value  
count          3000.00000  
mean         205846.27500  
std          113119.68747  
min           22500.00000  
25%          121200.00000  
50%          177650.00000  
75%          263975.00000  
max          500001.00000

### `shape`: データの行列数の取得

`shape` 属性は、(行数, 列数) のタプルを返します。

In [44]:
data.shape

(3000, 9)

`len` 関数を用いて、行数を確認することもできます。

In [45]:
len(data)

3000

## データの操作

### フィルタリング

`Series` や `DataFrame` では、特定の列を参照するだけでなく、条件に合致したデータのみを取得する、フィルタリングを行うことが可能です。
単一の条件によるフィルタリングは次のように行います。

```python
data[data["column"] > value]
```

In [46]:
data = pd.DataFrame({
    "product": ["Apple", "Banana", "Carrot", "Milk", "Beef"],
    "price": [120, 200, 150, 200, 500],
    "in_stock": [True, True, True, False, True],
    "category": ["fruit", "fruit", "vegetable", "dairy", "meat"]
})
data

product  price  in_stock   category
0   Apple    120      True      fruit
1  Banana    200      True      fruit
2  Carrot    150      True  vegetable
3    Milk    200     False      dairy
4    Beef    500      True       meat

価格が 150 円より高い商品を選択するには次のようにします。

In [47]:
data[data["price"] > 150]

product  price  in_stock category
1  Banana    200      True    fruit
3    Milk    200     False    dairy
4    Beef    500      True     meat

在庫がある商品データだけを取得する場合は次のようにします。

In [48]:
data[data["in_stock"]]

product  price  in_stock   category
0   Apple    120      True      fruit
1  Banana    200      True      fruit
2  Carrot    150      True  vegetable
4    Beef    500      True       meat

商品カテゴリーが果物のデータを選択してみます。

In [49]:
data[data["category"] == "fruit"]

product  price  in_stock category
0   Apple    120      True    fruit
1  Banana    200      True    fruit

### 新しい列の追加と削除

`DataFrame` には、新しいデータ列を追加したり、不要な列を削除したりする機能があります。
これにより、データの加工や整形が容易になります。

#### 新しい列の追加

新しい列を追加するには、以下のように列名を指定して値を代入します。

```python
data['new_column'] = value
```

In [50]:
data["price_with_tax"] = data["price"] * 1.1
data

product  price  in_stock   category  price_with_tax
0   Apple    120      True      fruit           132.0
1  Banana    200      True      fruit           220.0
2  Carrot    150      True  vegetable           165.0
3    Milk    200     False      dairy           220.0
4    Beef    500      True       meat           550.0

すべての商品に「日本」という原産国を表す列を追加する例も見てみましょう。

In [51]:
data["origin"] = "Japan"
data

product  price  in_stock   category  price_with_tax origin
0   Apple    120      True      fruit           132.0  Japan
1  Banana    200      True      fruit           220.0  Japan
2  Carrot    150      True  vegetable           165.0  Japan
3    Milk    200     False      dairy           220.0  Japan
4    Beef    500      True       meat           550.0  Japan

### 列の削除

列を削除するには、`drop` メソッドを使用し、`columns` キーワードで削除したい列名を指定します。

In [52]:
# 列の削除例
data.drop(columns="in_stock")

product  price   category  price_with_tax origin
0   Apple    120      fruit           132.0  Japan
1  Banana    200      fruit           220.0  Japan
2  Carrot    150  vegetable           165.0  Japan
3    Milk    200      dairy           220.0  Japan
4    Beef    500       meat           550.0  Japan

`drop` メソッドはデフォルトで新しい `DataFrame` を返しますが、元の `DataFrame` は変更しません。
元のデータに変更を適用する場合は、`inplace=True` を引数に追加します。

### 行の削除

特定の行を削除するには、行のインデックスを指定します。
以下の例では、インデックス `2` の行を削除します。

In [53]:
# 行の削除例
data.drop(2)

product  price  in_stock category  price_with_tax origin
0   Apple    120      True    fruit           132.0  Japan
1  Banana    200      True    fruit           220.0  Japan
3    Milk    200     False    dairy           220.0  Japan
4    Beef    500      True     meat           550.0  Japan

`axis` キーワードを使用することで、列 (`axis=1`) または行 (`axis=0`) を指定して削除することも可能です。
ただし、`columns` と `index` キーワードを使用する方法が直感的で推奨されます。

### 行列の入れ替え

`transpose` メソッドを使うと、行と列を入れ替えることができます。

これを、行列の転置といいます。

In [54]:
data.transpose()

0       1          2      3      4
product         Apple  Banana     Carrot   Milk   Beef
price             120     200        150    200    500
in_stock         True    True       True  False   True
category        fruit   fruit  vegetable  dairy   meat
price_with_tax  132.0   220.0      165.0  220.0  550.0
origin          Japan   Japan      Japan  Japan  Japan

`T` プロパティを使うことでも転置ができます。

In [55]:
data.T

0       1          2      3      4
product         Apple  Banana     Carrot   Milk   Beef
price             120     200        150    200    500
in_stock         True    True       True  False   True
category        fruit   fruit  vegetable  dairy   meat
price_with_tax  132.0   220.0      165.0  220.0  550.0
origin          Japan   Japan      Japan  Japan  Japan

## 基本的なデータの集計

データの集計は、データ分析において重要なステップの一つです。
`pandas` では、データをグループ化し、集計メソッドを適用することで、データセットの要約や分析を行うことができます。

### 集計メソッドの使用

`pandas` には、データを集計するためのいくつかのメソッドが用意されています。
これらのメソッドを使用することで、合計、平均、最大値、最小値などの計算が可能です。

- `.sum()`: 合計を計算
- `.mean()`: 平均を計算
- `.max()`: 最大値を計算
- `.min()`: 最小値を計算

これらのメソッドは、`DataFrame` の列に直接適用することも、`groupby` メソッドを使って特定のカテゴリごとに適用することもできます。
以下に例を示します。


In [56]:
# 全体の価格の平均を計算
data["price"].mean()

np.float64(234.0)

### グループ別集計の例

カテゴリごとに価格の平均を計算する例を示します。
まず、`groupby` メソッドを使ってカテゴリでグループ化します。

In [57]:
# グループ別の価格の平均を計算
data.groupby("category")["price"].mean()

category
dairy        200.0
fruit        160.0
meat         500.0
vegetable    150.0
Name: price, dtype: float64

この結果から、各カテゴリに属する商品の平均価格がわかります。
例えば、フルーツの平均価格は 160 円、肉類の平均価格は 500 円です。

このように、`pandas` の集計関数を使用することで、データセットの要約や傾向を把握することが可能です。

## データの確認・整理につかう属性・メソッド、関数

In [58]:
df = pd.DataFrame(
    [
        [1, 2, 3],
        [4, 5, 6],
    ]
)

`size` 属性は 行数 x 列数、つまりの要素の数を返します。

In [59]:
df.size

6

`values` 属性は `Series`, `DataFrame` の要素を `numpy` の `ndarray` というオブジェクトとして返します。

In [60]:
df.values

array([[1, 2, 3],
       [4, 5, 6]])

`Series` の [`value_counts`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html) メソッドは、データの値毎の頻度を数えてくれます。
簡単なデータのサマリーとして重宝します。

In [61]:
fruits = Series(
    ["Apple", "Banana", "Orange", "Banana", "Apple", "Orange", "Apple", "Banana"]
)
fruits

0     Apple
1    Banana
2    Orange
3    Banana
4     Apple
5    Orange
6     Apple
7    Banana
dtype: object

In [62]:
fruits.value_counts()

Apple     3
Banana    3
Orange    2
Name: count, dtype: int64

また、意図しないデータが紛れ込んでいないかを確認するのにも使えます。

In [63]:
Series(
    ["Apple", "Banana", "Orange", "Banana", "Apple", "Orange", "Apple", "-"]
).value_counts()

Apple     3
Banana    2
Orange    2
-         1
Name: count, dtype: int64